sched/signal/sig_dispatch.c

```diff
@@ -435,6 +473,14 @@ int nxsig_tcbdispatch(FAR struct tcb_s *stcb, siginfo_t *info)
 
   flags = enter_critical_section();
 
+  /* Make sure that there is always at least one sigpednq and sigq structure
+   * available, in case one needs to be queued later. Note that this breaks
+   * the critical section if it needs to allocate any new structures. So it
+   * needs to be done here before using the task state or sigprocmask.
+   */
+
+  flags = nxsig_alloc_dyn_pending(flags);
```

Comment:
move not move before line 474 and let nxsig_alloc_dyn_pending do the protection by self?

Reply:
We need to keep the critical section after adding the items to the free queues; otherwise it is in theory possible that other signal dispatcher thread in the same process steals it in between. And this one proceeds with 0 spare pending structures left. This is perhaps not the most common case, but still a race in theory ;)

Reply:
Ok